### PR TITLE
Inherited attributes cleanup migration

### PR DESCRIPTION
**Pull request by ajtmccarty**
more cubic comments than usual on this one b/c I initially included too many commits and the spec/plan changed halfway through to switch from retroactive timestamps to time-of-migration timestamps. also, it is big. I will have one more follow-up PR to this feature branch to fix up a few of cubic's suggestions that I've commented on below

# Why

When an existing node kind starts inheriting from a generic, older versions never created attribute rows for the newly inherited attributes on the kind's pre-existing nodes. Those nodes read the attribute back as unset, accepted update mutations that reported `ok: true` and persisted nothing, and never matched a filter on the attribute's value. PR 1 (#10113, in this PR's base) stopped new damage; installs damaged before it still carry it.

This PR repairs that accumulated damage at upgrade time.

Non-goals: it does not change how attributes are written at runtime, and it does not attempt to repair anything other than generic-inherited attributes — the only rows this damage shape can affect.

Closes IFC-2619

## What changed

**Behavioural**

- A new graph migration repairs every damaged (active node, generic-inherited attribute) pair. The default branch is repaired by the upgrade; every other branch is repaired by its post-upgrade rebase, which the upgrade already schedules by marking stale branches `NEED_UPGRADE_REBASE`.
- Repaired rows are created at the time the repair runs: attributes with a default carry it, attributes without one become editable rows holding no value, and NumberPool attributes get a reservation-aware allocation from the generic's pool, upserted if absent. Profile and template instances are repaired alongside concrete ones wherever they support the attribute.
- Graph migrations may now live in a package (`m{NNN}_{name}/`) rather than only a single module.

**Implementation notes**

- The migration is a `MigrationRequiringRebase`, which supplies the per-branch entry point that run-time timestamps require. An earlier design used retroactive timestamps so one upgrade pass could repair every branch; that was dropped, and the specs in this PR are updated to match (`AttributeAddQuery` loses the now-unused `write_at`).
- Repairs are chunked (`repair_batch_size`, default 1000) and idempotent, so a partial failure converges on rerun.
- The branch pass is scoped to kinds whose inherited attributes go beyond the default branch's schema, so it touches branch-origin damage only.
- Validation re-audits the default branch and fails the upgrade with per-kind detail if anything remains.

**Unchanged**: no schema changes, no API contract changes, and the runtime write path is untouched.

### Suggested review order

1. `backend/infrahub/core/migrations/graph/m076_heal_missing_attribute_rows/` — the migration itself
2. `backend/infrahub/core/migrations/graph/discovery.py` and `core/graph/__init__.py` — how it is picked up and versioned
3. `backend/infrahub/cli/db.py` and `core/branch/tasks.py` — the two independent fixes
4. Tests, then `dev/specs/` (if you hate yourself)

## How to review

The part worth the most scrutiny, according to Claude, is `_ensure_runtime_context` in `migration.py`. Invoked from the CLI, the migration arrives with no schema manager in the registry, so it builds one and has to recover the default branch's schema from the database. Getting this wrong is silent: the audit sees only core kinds, repairs nothing, and still reports success. The component suite drops the registry before running the migration (`drop_registry_schema()`) specifically so that path is exercised — without it the suite passes whether or not the recovery works.

`core/branch/tasks.py` is a small change with a subtle reason: rebasing moves the branch's `branched_from` forward, and the common-ancestor schema resolves against it, so the baselines have to be captured before the rebase rather than after.

## How to test

```bash
uv run pytest backend/tests/component/core/migrations/graph/m076_heal_missing_attribute_rows/
uv run pytest backend/tests/component/core/migrations/schema/test_node_kind_update.py
uv run pytest backend/tests/unit/core/migrations/test_discovery.py
```

Also validated end-to-end against a restored 5.1M-vertex dataset with damage created by the pre-fix code:

- `infrahub upgrade` repaired **57,042** rows across three kinds, matching the predicted count exactly, in 743 s. so it's not fast, but it keeps the user updated as it runs and doesn't take _too_ long.
- Per-flavour checks passed: schema defaults with `is_default: true`; no-default attributes as null-valued rows that accept an update and flip to `is_default: false`; 5,000 distinct in-range pool values with no duplicates, from a single `CoreNumberPool` registered against the generic.
- After rebasing a user branch, its own 4,000 rows were present and the nodes that existed only on that branch were repaired too.
- Re-running the migration reported `nothing to repair` and wrote nothing.

## Impact & rollout

- **Backward compatibility:** no breaking changes. The migration is one-shot and gated on `GRAPH_VERSION`; healthy installs write nothing.
- **Performance:** ~12 minutes on my machine for 57,042 rows on a 5.1M-vertex graph. Roughly 79% of that time went to 9% of the rows — the per-node NumberPool allocation path runs at about 2 minutes per 1,000 allocations, versus seconds for tens of thousands of chunked text rows. A bulk allocation path is deliberately deferred; the measured rate is the input for deciding when it becomes necessary.
- **Deployment notes:** safe to deploy. Branches are repaired when rebased, and the upgrade already marks them. Worth knowing that the upgrade output is now considerably more verbose by default, which is the point.

## Checklist

- [x] Tests added/updated
- [x] Changelog entry added
- [ ] External docs updated (if user-facing or ops-facing change)
- [x] Internal .md docs updated (internal knowledge and AI code tools knowledge)
- [x] I have reviewed AI generated content

---
## Summary by cubic
Backfills missing rows for newly inherited attributes so updates persist and filters match, and heals already damaged installs. Previously, nodes that gained generic-inherited attributes had no rows, updates returned ok but wrote nothing, and filters missed them; now rows are created and updates persist. Fixes IFC-2619.

- Forward fix: `NodeKindUpdateMigration` runs a `NodeAttributeAddMigration` for each newly inherited attribute (Profiles/Templates included). Applies defaults, allocates NumberPools, and writes branch-agnostic attributes on the global branch. Updates now return a non-null attribute id, set is_default: false, and persist.
- One-time heal: `m076_heal_missing_attribute_rows` repairs the default branch at upgrade; other branches repair on post-upgrade rebase. Idempotent and self-validating; allocates NumberPools and upserts the generic pool if missing. Operators should rebase long-lived branches after upgrade (the upgrade already schedules these).
- Infra: graph migrations can be packages (`m{NNN}_{name}/`); `GRAPH_VERSION` → 76; drop the unused `write_at` option from `AttributeAddQuery`.
- CLI/rebase: per-migration progress is always shown; `--verbose` only controls internal logger output. Resolve migration baselines before rebasing to capture the correct common-ancestor schema.

<sup>Written for commit a49ec12d6e98070f8b2060e25af4d5ea1eb547d2. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/opsmill/infrahub/pull/10224?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: backend/infrahub/cli/db.py [hunk 1]
```diff
@@ -130,7 +130,8 @@ async def do_migrate(
         root_node: The root node containing the current graph version.
```
[the remark this answers is not in the document]
Comment (ajtmccarty) [reply backend/infrahub/cli/db.py:130]: this is referencing an older possible fix so it is incorrect. and there's a PR description now

File: backend/tests/integration/schema_lifecycle/test_schema_add_inherited_generic.py [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+        refetched = await NodeManager.query(db=db, schema=SERVER_KIND, filters={"name__value": "server-1"})
+        assert len(refetched) == 1
+        status_attr = refetched[0].get_attribute(name="status")
+        assert status_attr.value == "planned"
+        assert status_attr.is_default is False
+
+    async def test_step02_filter_matches(self, db: InfrahubDatabase, initial_dataset: dict[str, str]) -> None:
+        # server-2 was never explicitly updated and must match on the inherited default
+        matches = await NodeManager.query(db=db, schema=SERVER_KIND, filters={"status__value": "active"})
+        assert [node.id for node in matches] == [initial_dataset["server2"]]
```
[the remark this answers is not in the document]
Comment (ajtmccarty) [reply backend/tests/integration/schema_lifecycle/test_schema_add_inherited_generic.py:179]: tests will always run in the same order, so doesn't really matter

File: backend/infrahub/core/migrations/query/attribute_add.py [hunk 3]
```diff
@@ -143,9 +137,6 @@ async def query_init(self, db: InfrahubDatabase, **kwargs: dict[str, Any]) -> No
             SET a.created_at = $current_time, a.created_by = $user_id, a.updated_at = $current_time, a.updated_by = $user_id
             SET n.updated_at = $current_time, n.updated_by = $user_id
         }
-        FOREACH (i in CASE WHEN has_attr_e.status = "deleted" THEN [1] ELSE [] END |
```
Comment (ajtmccarty) [inline backend/infrahub/core/migrations/query/attribute_add.py:146]: this was incorrectly adding a `to` time to a `status="deleted"` edge, which is something that we never do. once an edge is deleted, it is never going to be re-opened

File: backend/infrahub/core/migrations/schema/node_kind_update.py [hunk 4]
```diff
@@ -38,3 +48,38 @@ def get_nbr_migrations_executed(self) -> int:
...
+        migration_input: MigrationInput,
+        branch: Branch,
+        queries: Sequence[type[MigrationBaseQuery]] | None = None,
+    ) -> MigrationResult:
+        result = await super().execute(migration_input=migration_input, branch=branch, queries=queries)
+        if result.errors:
+            return result
+
+        # explicitly run migrations to add the new attributes to existing instances
+        for attribute in self._newly_inherited_attributes():
```
[the remark this answers is not in the document]
Comment (ajtmccarty) [reply backend/infrahub/core/migrations/schema/node_kind_update.py:63]: Confirmed. `_newly_inherited_attributes()` is a plain attribute-name diff with no check on `schema_path.field_name`, so a `node.name.update` or `node.namespace.update` that coincides with newly-gained inherited attributes does re-run the `force_inherited` sub-migrations. It is idempotent so it converges, but the redundant writes and the coupling into a re-labelling migration are both real.

This code came in with #10113, which is already merged, so we'll handle the gating in a follow-up PR.

File: backend/infrahub/core/migrations/schema/tasks.py [hunk 5]
```diff
@@ -71,6 +106,35 @@ async def schema_apply_migrations(message: SchemaApplyMigrationData) -> list[str
...
+
+    # kind-update migrations duplicate node vertices and re-point their edges; two concurrent
+    # duplications over overlapping vertices can each create a replacement vertex, so they must
+    # run one at a time
+    error_messages.extend(
+        await _apply_migration_batch(
+            message=message, migrations=kind_update_migrations, log=log, max_concurrent_execution=1
+        )
+    )
+    if error_messages:
```
[the remark this answers is not in the document]
Comment (ajtmccarty) [reply backend/infrahub/core/migrations/schema/tasks.py:129]: Confirmed. `backend/tests/unit/core/migrations/schema/test_tasks.py` covers only `split_migrations_by_phase` — nothing exercises `schema_apply_migrations` itself, so all three behaviors you list (kind-updates at `max_concurrent_execution=1`, phase ordering, and the phase-one error early-exit plus warning) are untested.

This flow came in with #10113, which is already merged, so we'll add the flow-level tests in a follow-up PR.

File: changelog/9284.fixed.md [hunk 6]
```diff
@@ -0,0 +1 @@
+Attributes gained when an existing node kind starts inheriting from a generic are now created on pre-existing nodes, including their Profile and Template instances, with NumberPool attributes allocated from the generic's pool. Schema migrations now run kind-update migrations to completion before all other migrations to prevent them from racing each other.
```
[the remark this answers is not in the document]
Comment (ajtmccarty) [reply changelog/9284.fixed.md:1]: Confirmed that both fragments will render into CHANGELOG.md.

Worth noting they describe different mechanisms — `9284.fixed.md` covers the forward path (new inheritance creating rows on pre-existing nodes, plus the serialized kind-update phase), while `+heal-missing-attribute-rows.fixed.md` covers the retroactive repair at upgrade. The first sentences do overlap heavily, though.

`9284.fixed.md` came in with #10113, which is already merged, and this PR only adds the heal fragment — so we'll reconcile the two in a follow-up PR.

File: dev/specs/ifc-2619-inherited-attr-migration/data-model.md [hunk 7]
```diff
@@ -35,45 +37,60 @@ Full structure per attribute (all created by the forward path and by healing):
...
+### Schema-graph vertices (read-only, historical)
 
-The schema's own graph representation records when schema elements changed:
+The schema's own graph representation records when schema elements changed. Inherited attributes are never persisted under the inheriting kind's schema vertex, so two owner-side timelines together bound the damage window: the generic's linkage to the `SchemaAttribute` vertex (when the generic gained the attribute) and the inheriting kind's `inherit_from` value edge (when the kind began inheriting the generic).
 
-- The **kind's schema node** and its `inherit_from` state → when the kind began inheriting the generic.
-- The **generic's schema node** and its attribute vertices → when the generic gained each attribute.
+> **Superseded.** The retroactive-timestamp design below was implemented and then dropped:
+> healed rows are created at run time, and branches receive their repairs when they rebase.
+> `AttributeAddQuery` no longer accepts a `write_at` argument. Kept for the derivation
```
[the remark this answers is not in the document]
Comment (ajtmccarty) [reply dev/specs/ifc-2619-inherited-attr-migration/data-model.md:46]: The contradiction is real, but it points the other way — `data-model.md` is the accurate side, so the suggested edit would introduce the error rather than fix it.

`AttributeAddQuery.__init__` currently takes `node_kinds`, `attribute_name`, `attribute_kind`, `branch_support`, `default_value` and `uuids`. There is no `write_at`; the write time comes from `self.at` in `query_init`. The base branch did carry a `write_at` parameter, and it was removed on this branch in 489e4004 — so "no longer accepts a `write_at` argument" is correct as written, and healed rows are stamped at run time rather than backdated.

The stale claim was the trailing clause of T030 in `tasks.md` ("`AttributeAddQuery` keeps its `uuids`/`write_at` parameters"), which described the plan at the time rather than the state it shipped in. Fixed in 50838ad5 by marking that clause superseded, which reconciles the two documents.

The same commit also rewrites the two edge-case bullets in `spec.md` flagged separately, since they described the same dropped retroactive-timestamp design.

File: backend/tests/component/core/migrations/graph/m076_heal_missing_attribute_rows/test_default_branch.py [hunk 8]
```diff
@@ -0,0 +1,525 @@
+import re
```
[the remark this answers is not in the document]
Comment (ajtmccarty) [reply backend/tests/component/core/migrations/graph/m076_heal_missing_attribute_rows/test_default_branch.py:1]: tests run in a defined order, so this is fine

File: dev/specs/ifc-2619-inherited-attr-migration/spec.md [hunk 9]
```diff
@@ -87,39 +90,39 @@ A branch user whose branch introduced the damaging schema change gets repaired d
...
-- **FR-007**: Healed NumberPool attribute rows MUST be created at migration-run time with a pool allocation that cannot collide with any existing reservation.
+- **FR-005**: The healing migration MUST create an active attribute row for every (active node, generic-inherited attribute) pair lacking one, on the default branch, regardless of how the row went missing. The audit scope is the attributes a kind inherits from its generics — the only rows this damage shape can affect — discovered from the persisted schema graph.
+- **FR-006**: Healed rows (other than NumberPool) MUST carry the schema default value, or no value when the attribute has no default, and MUST be created at the time the repair runs. *(Superseded 2026-08-04: this requirement originally mandated a retroactive timestamp derived from the schema graph, which is why branch repair no longer happens at upgrade time — see FR-009.)*
```
[the remark this answers is not in the document]
Comment (ajtmccarty) [reply dev/specs/ifc-2619-inherited-attr-migration/spec.md:94]: updated

File: backend/infrahub/core/migrations/graph/m076_heal_missing_attribute_rows/queries.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+        self.add_to_query(query)
+        self.return_labels = ["node_uuid", "attribute_name"]
+
+    def get_data(self) -> Generator[AttributeHealDetectionQueryResult, None, None]:
+        for result in self.get_results():
+            yield AttributeHealDetectionQueryResult(
+                node_uuid=result.get_as_type("node_uuid", str),
+                attribute_name=result.get_as_type("attribute_name", str),
+            )
```
Comment (polmichel) [inline backend/infrahub/core/migrations/graph/m076_heal_missing_attribute_rows/queries.py:83]: This pattern is clean in my opinion

File: backend/infrahub/core/migrations/graph/m076_heal_missing_attribute_rows/queries.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+        UNWIND $attribute_names AS attr_name
+        CALL (n, attr_name) {
+            OPTIONAL MATCH (n)-[r:HAS_ATTRIBUTE]-(a:Attribute { name: attr_name })
+            WHERE %(branch_filter)s
+            RETURN r AS has_attr_e
+            ORDER BY r.branch_level DESC, r.from DESC, r.status ASC
+            LIMIT 1
+        }
+        WITH n, attr_name, has_attr_e
+        WHERE has_attr_e IS NULL OR has_attr_e.status = "deleted"
```
Comment (polmichel) [inline backend/infrahub/core/migrations/graph/m076_heal_missing_attribute_rows/queries.py:68]: Just to be sure to understand this line, which seems to refer to _"A node whose only row for an attribute is
    tombstoned counts as damaged"_ inside the class description: 
This is to illustrate the case where some operation has deleted this attribute, and we were supposed to have a new one which had not been created for whatever reason.

Is that a case you've seen during the fix?

File: backend/infrahub/core/migrations/graph/m076_heal_missing_attribute_rows/migration.py [hunk 12]
```diff
@@ -0,0 +1,490 @@
...
+        return created
+
+    async def _repair_pool_backed_attribute(
+        self,
+        migration_input: MigrationInput,
+        branch: Branch,
+        node_schema: NodeSchema,
+        attribute: AttributeSchema,
+        node_uuids: set[str],
+    ) -> int:
```
Comment (polmichel) [inline backend/infrahub/core/migrations/graph/m076_heal_missing_attribute_rows/migration.py:408]: There were situations where nodes could be "orphaned," in the sense that they were not attached to the generic schema. I don't think this PR is meant to fix that, but have you considered the cases where pool-backed attributes already have values locked in by those "orphaned" nodes?

**Review by polmichel — APPROVED** [review]
LGTM the test suite is really clean. I only got two questions regarding one edge case and some details about the internal behavior of a part of this development

File: backend/tests/component/core/migrations/graph/m076_heal_missing_attribute_rows/test_number_pool.py [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+        nodes = await NodeManager.get_many(
+            db=db,
+            branch=default_branch_scope_class,
+            ids=[*seeded.damaged_uuids, seeded.runtime_server_uuid],
+            include_metadata=MetadataQueryOptions(attribute_level=MetadataOptions.SOURCE),
+        )
+        rack_units = {node_id: node.get_attribute(name="rack_unit").value for node_id, node in nodes.items()}
+        assert set(rack_units) == {*seeded.damaged_uuids, seeded.runtime_server_uuid}
+        assert all(isinstance(value, int) and 1 <= value <= 100 for value in rack_units.values())
+        assert len(set(rack_units.values())) == 4
```
Comment (polmichel) [inline backend/tests/component/core/migrations/graph/m076_heal_missing_attribute_rows/test_number_pool.py:126]: Can we make these assertions stricter? 
Like
```
  assert rack_units[seeded.runtime_server_uuid] == 1
  assert {rack_units[uuid] for uuid in seeded.damaged_uuids} == {2, 3, 4}
```

File: backend/infrahub/core/migrations/graph/m076_heal_missing_attribute_rows/queries.py (hunk 11 above)
Comment (ajtmccarty) [reply backend/infrahub/core/migrations/graph/m076_heal_missing_attribute_rows/queries.py:68]: yes that's exactly right. this would be the case where a user wants to move an attribute from a node schema up to a generic schema, so they 
1. delete the attribute from a node schema
2. add the attribute to a generic schema
3. add the generic schema to the node schema's inherit_from

depending on whether steps 2 and 3 happen at the same time or not, we might handle it correctly in the current code, but we can't be sure

File: backend/infrahub/core/migrations/graph/m076_heal_missing_attribute_rows/migration.py (hunk 12 above)
Comment (ajtmccarty) [reply backend/infrahub/core/migrations/graph/m076_heal_missing_attribute_rows/migration.py:408]: I don't think I understand what the issue is that you're describing. this migration should will repair any NumberPool attributes that are inherited from a generic and missing values, including unlinked/orphaned attributes. I guess I'm not sure what an "orphaned" node is in this situation b/c we only link nodes to their generic schema using labels at the database level, ie `:TestDog:TestAnimal`

File: backend/tests/component/core/migrations/graph/m076_heal_missing_attribute_rows/test_number_pool.py (hunk 13 above)
Comment (ajtmccarty) [reply backend/tests/component/core/migrations/graph/m076_heal_missing_attribute_rows/test_number_pool.py:126]: good suggestion. added
